### PR TITLE
Update IP assignment timestamps and EC2 handling

### PR DESCRIPTION
**Pull request by ashleyzhang3**
- ELBs and their ENIs now use CreatedTime as their IP assignment timestamps
- ENIs that belong to an ELB but arriving alone now use AttachTime as their IP assignment timestamps
- EC2 CREATE events involving multiple ENIs are now split into multiple events, with each ENI's AttachTime as the IP assignment timestamp
- DELETE events are unchanged; they still use ConfigurationItemCaptureTime

### CODE REVIEW

**Comment by ashleyzhang3** [comment]
The Travis builds are failing because of a couple tests on transforming ELBs: `TestTransformELB/elb-updated` and `TestTransformELB/elbv2-updated`. These two tests are trying to compare the `Output` structs returned by the `Handle` function, and there are differences between the expected and actual in the pointer address of one of the fields inside. This causes the test and therefore build to fail. The master branch circumvents this by comparing the expected output with itself on line 218 of `elb_test.go`, effectively a no-op. I replaced this behavior with a true expected/actual check; please let me know if I should put it back, or make any other changes in order to merge.

`elb_test.go:218: 
        	Error Trace:	elb_test.go:218
        	Error:      	Not equal: 
        	            	expected: []v1.Change{v1.Change{PublicIPAddresses:[]string(nil), PrivateIPAddresses:[]string(nil), Hostnames:[]string(nil), RelatedResources:[]string(nil), TagChanges:[]v1.TagChange{v1.TagChange{UpdatedValue:(*v1.Tag)(0xc02fa0), PreviousValue:(*v1.Tag)(nil)}}, ChangeType:"ADDED"}}
        	            	actual  : []v1.Change{v1.Change{PublicIPAddresses:[]string(nil), PrivateIPAddresses:[]string(nil), Hostnames:[]string(nil), RelatedResources:[]string(nil), TagChanges:[]v1.TagChange{v1.TagChange{UpdatedValue:(*v1.Tag)(0xc000124280), PreviousValue:(*v1.Tag)(nil)}}, ChangeType:"ADDED"}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -9,4 +9,4 @@
        	            	     UpdatedValue: (*v1.Tag)({
        	            	-     Key: (string) (len=4) "key1",
        	            	-     Value: (string) (len=1) "1"
        	            	+     Key: (string) (len=4) "key2",
        	            	+     Value: (string) (len=1) "2"
        	            	     }),
        	Test:       	TestTransformELB/elb-updated
`
This is the unit test output for both this new branch and for master, when using the expected/actual comparison.

File: pkg/handlers/v1/ec2.go [hunk 1]
```diff
@@ -69,40 +78,114 @@ type ec2ChangedPropsARN struct {
...
+func initializeAttachMapEntry(attachMap map[string]map[string][]string, attachTime string) {
+	attachMap[attachTime] = make(map[string][]string)
+	attachMap[attachTime][private] = make([]string, 0)
+	attachMap[attachTime][public] = make([]string, 0)
+	attachMap[attachTime][dns] = make([]string, 0)
+}
+
+func separateAddedChanges(baseOutput Output, addedChange Change, privateIPMap map[string]string, publicIPMap map[string]string, hostnameMap map[string]string) []Output {
+	outputs := make([]Output, 0)
+	// map of attachTimes --> map{
```
Comment (gcase555) [inline pkg/handlers/v1/ec2.go:125]: question: do we intend do remove this commented out code?

**Comment by gcase555** [comment]
just FYI @ashleyzhang3 I made this PR https://github.com/asecurityteam/awsconfig-transformerd/pull/24 to fix broken linting(you have some of these changes here as well though but once I merge it, you should only see your needed changes here)

File: pkg/handlers/v1/ec2.go (hunk 1 above)
Comment (ashleyzhang3) [reply pkg/handlers/v1/ec2.go:125]: This is a comment to explain the structure of attachMap, I will update it to make it more explanatory

File: pkg/handlers/v1/ec2.go [hunk 2]
```diff
@@ -69,40 +78,113 @@ type ec2ChangedPropsARN struct {
...
+		return []Output{}, err
 	}
-	change := extractEC2NetworkInfo(&config)
-	change.ChangeType = added
-	output.Changes = append(output.Changes, change)
-	return output, nil
+
+	hasEni := false
+	for _, ni := range config.NetworkInterfaces {
+		ni := ni
```
Comment (akhalina) [inline pkg/handlers/v1/ec2.go:96]: Is there a reason for this line? It seems repetitive here.

File: pkg/handlers/v1/ec2.go [hunk 3]
```diff
@@ -69,40 +78,113 @@ type ec2ChangedPropsARN struct {
...
+	attachMap[attachTime][dns] = make([]string, 0)
+}
+
+func separateAddedChanges(baseOutput Output, addedChange Change, privateIPMap map[string]string, publicIPMap map[string]string, hostnameMap map[string]string) []Output {
+	outputs := make([]Output, 0)
+	// attachMap is a map of attachTimes (string) to a map with the following pairs:
+	// 	"privateIPs": array of private IPs
+	// 	"publicIPs": array of public IPs
+	// 	"hostnames": array of hostnames
+	attachMap := make(map[string]map[string][]string)
```
Comment (akhalina) [inline pkg/handlers/v1/ec2.go:129]: I think if you're at a point where you need a complex comment like this then it may be beneficial to create a struct instead. I think nested maps is a bit hard to follow.

File: pkg/handlers/v1/ec2.go [hunk 4]
```diff
@@ -114,24 +196,39 @@ func (t ec2Transformer) Update(event awsConfigEvent) (Output, error) {
...
+		deletedOutput := baseOutput
+		deletedOutput.Changes = append(deletedOutput.Changes, deletedChange)
+		outputs = append(outputs, deletedOutput)
 	}
-	return output, nil
+	return outputs, nil
 }
 
-func (t ec2Transformer) Delete(event awsConfigEvent) (Output, error) {
+func (t ec2Transformer) Delete(event awsConfigEvent) ([]Output, error) {
```
Comment (akhalina) [inline pkg/handlers/v1/ec2.go:228]: Note for reviewers: We are still using configurationItemCaptureTime for the deletion events because we don't currently have any other options. That's why this method isn't changing much.

File: pkg/handlers/v1/elb.go [hunk 5]
```diff
@@ -28,68 +35,97 @@ func extractELBNetworkInfo(config *elbConfiguration) Change {
 
 type elbTransformer struct{}
 
-func (t elbTransformer) Create(event awsConfigEvent) (Output, error) {
+// Some ELBs have createdTime in the form of milliseconds, which is int64 in Go,
```
Comment (akhalina) [inline pkg/handlers/v1/elb.go:38]: Nit: It looks like this is a difference in load balancers, so classic load balancers seem to have the millisecond timestamp and the rest have a string. I think I would prefer to handle the timestamp based on that. I'm not super into this raw message stuff, so I'd prefer to avoid it.

File: pkg/handlers/v1/ec2.go [hunk 6]
```diff
@@ -69,40 +78,113 @@ type ec2ChangedPropsARN struct {
...
-func (t ec2Transformer) Update(event awsConfigEvent) (Output, error) {
-	output, err := getBaseOutput(event.ConfigurationItem)
+func initializeAttachMapEntry(attachMap map[string]map[string][]string, attachTime string) {
+	attachMap[attachTime] = make(map[string][]string)
+	attachMap[attachTime][private] = make([]string, 0)
+	attachMap[attachTime][public] = make([]string, 0)
+	attachMap[attachTime][dns] = make([]string, 0)
+}
+
+func separateAddedChanges(baseOutput Output, addedChange Change, privateIPMap map[string]string, publicIPMap map[string]string, hostnameMap map[string]string) []Output {
```
Comment (akhalina) [inline pkg/handlers/v1/ec2.go:123]: Nit: I think it would be better to do this at the ENI level. It may require a bit of refactoring to the "symmetricDifference" function, but I think that is more fool proof and easier to read and reason about. Then you won't need such a complicated method that almost reconstructs the ENI.

File: pkg/handlers/v1/ec2.go (hunk 2 above)
Comment (ashleyzhang3) [reply pkg/handlers/v1/ec2.go:96]: That was to prevent a lint error that disallows making a reference of a for loop value (&ni)
- [ ] Change this to use the array index instead

File: pkg/handlers/v1/ec2.go (hunk 3 above)
Comment (ashleyzhang3) [reply pkg/handlers/v1/ec2.go:129]: - [ ] Convert map-of-maps into dedicated struct

File: pkg/handlers/v1/elb.go (hunk 5 above)
Comment (ashleyzhang3) [reply pkg/handlers/v1/elb.go:38]: - [ ] Unmarshal events by load balancer type, instead of raw message parsing

File: pkg/handlers/v1/ec2.go (hunk 6 above)
Comment (akhalina) [reply pkg/handlers/v1/ec2.go:123]: Ashley and I chatted about this, there could be some weird edge cases that complicate my suggested method a little. Ex: The same IP was deleted via one ENI and reattached via another. While what I suggested could still work, it would result in likely a 3 tiered for loop which may be more complicated than what this is doing. So I'm okay with keeping this as is.

File: pkg/handlers/v1/ec2.go (hunk 1 above)
Comment (ashleyzhang3) [reply pkg/handlers/v1/ec2.go]: This comment is no longer necessary due to refactoring this into a simpler struct, removed

**Comment by gcase555** [comment]
suggestion: notably the test coverage has gone down by 2.3%(as per codecov), are there more tests we could potentially add to get this number higher? transformerd is one of the most important pieces of our asset inventory data pipeline, trying to keep our coverage high(especially when making large changes) should generally be desirable.

Is it possible we can try to get the coverage back to 90%? are there more tests we can add without exploding the scope of this change too much?

File: pkg/handlers/v1/eni_test.go [hunk 7]
```diff
@@ -24,7 +24,7 @@ func TestTransformENI(t *testing.T) {
 			InputFile: "eni.create.json",
 			ExpectedOutput: Output{
 				AccountID:    "123456789",
-				ChangeTime:   "2020-08-21T12:00:00.000Z",
+				ChangeTime:   "2020-08-14T21:00:00.000Z",
```
Comment (gcase555) [inline pkg/handlers/v1/eni_test.go:27]: question: why did we need to change the time here? did the our changes here effect how this test would behave?

Comment (ashleyzhang3) [reply pkg/handlers/v1/eni_test.go:27]: The timestamp was changed to be the ENI's attachTime, instead of the ConfigurationItemCaptureTime, as part of the shift away from using the capture time.

**Comment by ashleyzhang3** [comment]
> suggestion: notably the test coverage has gone down by 2.3%(as per codecov), are there more tests we could potentially add to get this number higher? transformerd is one of the most important pieces of our asset inventory data pipeline, trying to keep our coverage high(especially when making large changes) is generally desirable.
- [ ] Add test cases to increase coverage (elb.go)

**Comment by gcase555** [comment]
That test coverage increase

File: pkg/handlers/v1/transformer_test.go [hunk 8]
```diff
@@ -273,17 +343,78 @@ func TestTransformEC2(t *testing.T) {
...
+				assert.Equal(t, tt.ExpectedOutput[i].ARN, output[i].ARN)
+				assert.Equal(t, tt.ExpectedOutput[i].Tags, output[i].Tags)
+				assert.Equal(t, tt.ExpectedOutput[i].ChangeTime, output[i].ChangeTime)
+				assert.Equal(t, tt.ExpectedOutput[i].Changes, output[i].Changes)
+			}
+		})
+	}
+}
+
+func TestTransformEC2Errors(t *testing.T) {
```
Comment (gcase555) [inline pkg/handlers/v1/transformer_test.go:364]: table based tests ⭐

**Review by gcase555 — APPROVED** [review]

File: pkg/handlers/v1/ec2.go [hunk 9]
```diff
@@ -67,42 +70,115 @@ type ec2ChangedPropsARN struct {
...
+	}
+	for _, ip := range addedChange.PublicIPAddresses {
+		attachTime := publicIPMap[ip]
+		if _, ok := attachMap[attachTime]; !ok {
+			initializeAttachMapEntry(attachMap, attachTime)
+		}
+		attachMap[attachTime].PublicIPs = append(attachMap[attachTime].PublicIPs, ip)
+	}
+	for _, hostname := range addedChange.Hostnames {
+		// the attachTime will always already be in the map, because adding the private IP(s)
```
Comment (alecbert-atl) [inline pkg/handlers/v1/ec2.go:141]: Nit on this comment: wouldn't it be because the public IPs have created an entry?

File: pkg/handlers/v1/elb.go [hunk 10]
```diff
@@ -28,68 +35,104 @@ func extractELBNetworkInfo(config *elbConfiguration) Change {
 
 type elbTransformer struct{}
 
-func (t elbTransformer) Create(event awsConfigEvent) (Output, error) {
+// Classic ELBs have createdTime in the form of milliseconds, which is int64 in Go,
```
Comment (alecbert-atl) [inline pkg/handlers/v1/elb.go:38]: I appreciate these comments on why we're doing this a lot. Pieces of domain knowledge like this tend to get forgotten months out from when we first write code and become a pain to relearn when we have to touch the codebase later

File: pkg/handlers/v1/transformer_test.go [hunk 11]
```diff
@@ -121,18 +121,18 @@ func TestTransformEC2(t *testing.T) {
 				},
 				Changes: []Change{
 					{
-						PrivateIPAddresses: []string{},
-						PublicIPAddresses:  []string{"34.219.72.29"},
-						Hostnames:          []string{"ec2-34-219-72-29.us-west-2.compute.amazonaws.com"},
-						ChangeType:         "ADDED",
+						// PrivateIPAddresses: []string{},
```
Comment (alecbert-atl) [inline pkg/handlers/v1/transformer_test.go:124]: Question: I think I missed how this happens when I was reading through. What changed that made us omit the PrivateIPAddresses array when it's empty (to be clear, I think that's an improvement!)

**Review by alecbert-atl — COMMENTED** [review]
Super small nit and then one clarifying question. I'd like to understand the question before I hit approve, but I've got no problems with the PR from what I've seen.

Great work!

File: pkg/handlers/v1/transformer_test.go (hunk 11 above)
Comment (gcase555) [reply pkg/handlers/v1/transformer_test.go:124]: Also its probably worth deleting commented out code if were not going to use it.

File: pkg/handlers/v1/ec2.go (hunk 9 above)
Comment (ashleyzhang3) [reply pkg/handlers/v1/ec2.go:141]: The map key is the attachTime, and the private IP is the first to be added to a map entry, so it will create an entry with that attachTime. Based on the sample data I've seen, instances always have a private IP - although I'm not sure if that's always true, since I've only seen a small number of samples. The function `extractNetworkInterfaceInfo` seems to assume that there is at least 1 private IP, otherwise it returns no IPs or hostnames altogether.

Comment (alecbert-atl) [reply pkg/handlers/v1/ec2.go:141]: 👍 I like that. Even after this amount of time I get thrown by corner cases on config events, so extra safety is always appreciated.

File: pkg/handlers/v1/transformer_test.go (hunk 11 above)
Comment (ashleyzhang3) [reply pkg/handlers/v1/transformer_test.go:124]: The PrivateIPAddresses array is now nil if there are no items, instead of an empty array.
- [ ] Delete commented out code

**Review by alecbert-atl — APPROVED** [review]

**Review by gcase555 — APPROVED** [review]
